codegen/generate_test.go

```diff
@@ -0,0 +1,92 @@
+package codegen
+
+import (
+	"testing"
+)
+
+// testPaths is a sampling of real paths that
+// originate in the OpenAPI doc, pulled via:
+// $ cat testdata/openapi.json | jq '.paths' | jq 'keys[]'
+var testPaths = []string{
+	"/database/roles",
+	"/database/roles/{name}",
+	"/auth/userpass/users/{username}/password",
+	"/auth/userpass/users/{username}/policies",
+	"/transit/export/{type}/{name}/{version}",
+}
+
+func TestCodeFilePath(t *testing.T) {
+	expected := []struct {
+		DataSourceFilePath string
+		ResourceFilePath   string
+	}{
+		{
+			DataSourceFilePath: "/generated/datasources/database/roles.go",
+			ResourceFilePath:   "/generated/resources/database/roles.go",
+		},
+		{
+			DataSourceFilePath: "/generated/datasources/database/roles/name.go",
+			ResourceFilePath:   "/generated/resources/database/roles/name.go",
+		},
+		{
+			DataSourceFilePath: "/generated/datasources/auth/userpass/users/username/password.go",
+			ResourceFilePath:   "/generated/resources/auth/userpass/users/username/password.go",
```

Comment:
This seems very verbose in naming. `/generated/resources/auth/userpass` seems very reasonable, but curious of the benefit of going deeper into `user/username/`. Looking at the [API docs][userpass] I more likely expect a `.go` file for each endpoint under `username`, e.g.  `/generated/resources/auth/userpass/user.go` and `/generated/resources/auth/userpass/login.go`. What do you think about reducing the nesting here?


[userpass]: https://www.vaultproject.io/api-docs/auth/userpass

Reply:
Maybe, I'm not entirely certain the way you're thinking of. 

Here's an example of all the userpass endpoints, with the .go file that currently gets generated for it:
```
/auth/userpass/duo/access =>                /generated/resources/auth/userpass/duo/access.go
/auth/userpass/duo/config =>                /generated/resources/auth/userpass/duo/config.go
/auth/userpass/login/{username} =>          /generated/resources/auth/userpass/login/username.go
/auth/userpass/mfa_config =>                /generated/resources/auth/userpass/mfa_config.go
/auth/userpass/users =>                     /generated/resources/auth/userpass/users.go
/auth/userpass/users/{username} =>          /generated/resources/auth/userpass/users/username.go
/auth/userpass/users/{username}/password => /generated/resources/auth/userpass/users/username/password.go
/auth/userpass/users/{username}/policies => /generated/resources/auth/userpass/users/username/policies.go
```
Can you copy pasta that and put what the result would be the way you're thinking of?

Reply:
I was thinking something like this would be simpler:

```
/auth/userpass/duo/access =>                /generated/resources/auth/userpass/duo.go
/auth/userpass/duo/config =>                /generated/resources/auth/userpass/duo.go
/auth/userpass/login/{username} =>          /generated/resources/auth/userpass/login.go
/auth/userpass/mfa_config =>                /generated/resources/auth/userpass/mfa_config.go
/auth/userpass/users =>                     /generated/resources/auth/userpass/users.go
/auth/userpass/users/{username} =>          /generated/resources/auth/userpass/users.go
/auth/userpass/users/{username}/password => /generated/resources/auth/userpass/users.go
/auth/userpass/users/{username}/policies => /generated/resources/auth/userpass/users.go
```